### PR TITLE
feat: Integrate NFT Key Marketplace (Avalanche) with DappRadar

### PR DESCRIPTION
**Pull request by westonnelson**
NFT Key - Avalanche NFT Marketplace
Smart Contract Integration for DappRadar

Submission by: Weston Nelson
Twitter: https://twitter.com/westonnelson
Discord: Weston Nelson#2021
ENS: westonnelson.eth

### CODE REVIEW

File: adapters/nft-key/index.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+const Ethereum = require("../../sdk/EVMC");
+const axios = require("axios");
+const URL = "http://nft-sales-service.dappradar.com/open-source";
+const KEY = process.env.DAPPRADAR_API_KEY;
+const path = require("path");
+
+class NFT_K {
+    // stands for NFT Key
+    constructor() {
+        this.name = "NFTK";
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:14]: Class name and provider name should remain the same as folder name.

File: adapters/nft-key/index.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
+require("dotenv").config();
+
+const moment = require("moment");
+const BigNumber = require("bignumber.js");
+const Ethereum = require("../../sdk/EVMC");
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:5]: Use `avalanche.js` SDK for Avalanche protocol.

File: adapters/nft-key/index.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+const URL = "http://nft-sales-service.dappradar.com/open-source";
+const KEY = process.env.DAPPRADAR_API_KEY;
+const path = require("path");
+
+class NFT_K {
+    // stands for NFT Key
+    constructor() {
+        this.name = "NFTK";
+        this.symbol = null;
+        this.token = "0x0000000000000000000000000000000000000000";
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:16]: For native AVAX token use `avax` instead of `0x0`.

File: adapters/nft-key/index.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        this.protocol = "AVAILABLE_PROTOCOLS.AVALANCHE";
+        this.block = 12855192;
+        this.contract = "0x1a7d6ed890b6c284271ad27e7abe8fb5211d0739";
+        this.events = struct Listing {
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:20]: `this.events` array should hold only names of events that should be tracked.
E.g. `this.events = ["TokenBought"]`

Comment (liutaurasjanuskas) [reply adapters/nft-key/index.js:20]: Also, do not copy/paste code from Solidity to Javascript :)

File: adapters/nft-key/index.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        );
+        return resp.data;
+    };
+    stop = async () => {
+        this.sdk.stop();
+    };
+
+    getBuyer = async event => {
+        const buyer = event.returnValues.owner;
+        if (event.event === "Offer, OfferCreated, OfferCancelled, OfferAccepted") {
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:116]: It is impossible to get this condition as `true`.

File: adapters/nft-key/index.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            return txReceipt.from;
+        }
+        return buyer;
+    };
+
+    process = async event => {
+        const block = await this.sdk.getBlock(event.blockNumber);
+        const timestamp = moment.unix(block.timestamp).utc();
+        const po = await this.getPrice(block.timestamp);
+        const nativePrice = new BigNumber(event.returnValues.cost).dividedBy(10 ** this.symbol.decimals);
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:130]: Your event doesn't have `cost` property.

File: adapters/nft-key/index.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        const buyer = await this.getBuyer(event);
+        if (!buyer) {
+            return;
+        }
+
+        const labelHash = event.returnValues.label;
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:136]: Your event doesn't have `label` property.

File: adapters/nft-key/index.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            provider_contract: this.contract, // the providers contract from which you get data
+            protocol: this.protocol,
+            nft_contract: this.contract,
+            nft_id: tokenId,
+            token: this.token,
+            token_symbol: this.symbol.symbol,
+            amount: 1,
+            price: nativePrice.toNumber(),
+            price_usd: nativePrice.multipliedBy(po.price).toNumber(),
+            seller: this.contract, // Exchange happens on NFT Key
```
Comment (liutaurasjanuskas) [inline adapters/nft-key/index.js:149]: I'm pretty sure NFT Key marketplace is not a seller.

**Review by liutaurasjanuskas — CHANGES_REQUESTED** [review]
Fix the issues that I've mentioned. Thanks

**Comment by liutaurasjanuskas** [comment]
We've already merged another PR of this provider: https://github.com/dappradar/nft-sales-adapters/commit/d161f8e13998447f9e7c18fcdc428c3f26fea5cc

File: adapters/nft-key/index.js (hunk 2 above)
Comment (westonnelson) [reply adapters/nft-key/index.js:5]: I see - got it.

File: adapters/nft-key/index.js (hunk 3 above)
Comment (westonnelson) [reply adapters/nft-key/index.js:16]: Was a bit confused on what was being asked here, but can update this as well, and then I assumed for all chains would be the same, ie ftm, bnb, matic, etc.

File: adapters/nft-key/index.js (hunk 4 above)
Comment (westonnelson) [reply adapters/nft-key/index.js:20]: Got it, thank you! I used the NFTKey Docs and what they provided as Events Emitted from Marketplace here.

File: adapters/nft-key/index.js (hunk 6 above)
Comment (westonnelson) [reply adapters/nft-key/index.js:130]: Thank you for pointing this out.

File: adapters/nft-key/index.js (hunk 8 above)
Comment (westonnelson) [reply adapters/nft-key/index.js:149]: They provide 2 different sets of Events Emitted in their Docs...perhaps the Global Marketplace is the correct set to use I would think then. 

This is where I found the Events:
https://docs.nftkey.app/faqs/

**Review by westonnelson — COMMENTED** [review]
Thanks for the feedback! @liutaurasjanuskas

I saw the reference to the PR regarding AVAX and NFTKey merged, so thank you for the awareness on that. Cheers!

**Comment by liutaurasjanuskas** [comment]
You're always welcome to contribute with other marketplaces!

**Comment by westonnelson** [comment]
Thank you! I was going to ask that as I was researching; what would be helpful? I am looking into the Kalao Marketplace on Avalanche and think I could do so with all your helpful feedback but their smart contract isn't verified on SnowTrace so working on how to get the ABI here. There does seem to be quite a bit more activity than I thought would be going on here though so happy to try this one, or if you have any others you would prefer let me know! 

-Weston

https://snowtrace.io/address/0x11AC3118309A7215c6d87c7C396e2DF333Ae3A9C

**Pull request closed without merging**